components/feature/webnotifications/src/main/java/mozilla/components/feature/webnotifications/WebNotificationFeature.kt

```diff
@@ -0,0 +1,103 @@
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.feature.webnotifications
+
+import android.app.NotificationChannel
+import android.app.NotificationManager
+import android.content.Context
+import android.os.Build
+import androidx.core.app.NotificationCompat
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.GlobalScope
+import kotlinx.coroutines.launch
+import mozilla.components.browser.icons.BrowserIcons
+import mozilla.components.concept.engine.Engine
+import mozilla.components.concept.engine.webnotifications.WebNotification
+import mozilla.components.concept.engine.webnotifications.WebNotificationDelegate
+import mozilla.components.support.base.log.logger.Logger
+import java.lang.UnsupportedOperationException
+
+private const val NOTIFICATION_CHANNEL_ID = "mozac.feature.webnotifications.generic"
+
+/**
+ * Feature implementation for configuring and displaying web notifications to the user.
+ *
+ * Initialize this feature globally once on app start
+ * ```Kotlin
+ * WebNotificationFeature(
+ *     applicationContext, engine, icons, R.mipmap.ic_launcher, BrowserActivity::class.java
+ * )
+ * ```
+ *
+ * @param context The application Context.
+ * @param engine The browser engine.
+ * @param browserIcons The entry point for loading the large icon for the notification.
+ * @param launcherIcon The small icon for the notification.
+ * @param activityClass The Activity that the notification will launch if user taps on it
+ */
+class WebNotificationFeature(
+    val context: Context,
+    val engine: Engine,
+    val browserIcons: BrowserIcons,
+    val launcherIcon: Int,
+    val activityClass: Class<*>?
+) {
+    private val logger = Logger("AutoPushFeature")
+    private var pendingRequestId = 0
+    private var notificationId = 0
+    private val notificationIdMap = HashMap<String, Int>()
+
+    init {
+        try {
+            engine.registerWebNotificationDelegate(FeatureWebNotificationDelegate())
+        } catch (e: UnsupportedOperationException) {
+            logger.error("failed to register for web notification delegate", e)
+        }
+    }
+
+    private inner class FeatureWebNotificationDelegate : WebNotificationDelegate {
+        override fun onShowNotification(webNotification: WebNotification) {
+            val notificationManager = context.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
+            val nativeNotificationBridge = NativeNotificationBridge(browserIcons)
+            val channelId = ensureChannelExists(context, notificationManager)
+
+            notificationIdMap[webNotification.tag]?.let {
+                notificationManager.cancel(it)
+            }
+
+            notificationIdMap[webNotification.tag] = notificationId
+            GlobalScope.launch(Dispatchers.IO) {
+                notificationManager.notify(notificationId++, nativeNotificationBridge.convertToAndroidNotification(
```

Comment:
I'm not a fan of inline `++` because it's hard to remember if the return value is before or after the addition. I think it'd be better to store the desired IDs in a local variable separate from incrementing.

Reply:
Sure, I'll move the increment to the end of the block.